Connect Pool DB with Api

Anviking opened the pull request:
# Issue Number

#713 

# Overview

- [x] I have implemented a `GET /stake-pool` that reads and combines metrics from `Pool.DB`
- [x] TODO: Figure out when we are unsynced
- [ ] TODO: Look for things to test if possible in this PR

# Comments

### Self-node with a very short epoch-length (results are unavailable across epoch boundaries)
![self-node-short-epochs](https://user-images.githubusercontent.com/304423/67398367-d0edc680-f5aa-11e9-86da-ac13337e2a63.gif)

### Testnet syncing which is interrupted and resumed
![testnet-2](https://user-images.githubusercontent.com/304423/67399072-f7f8c800-f5ab-11e9-9fb1-38e8886aeba4.gif)


<!-- Additional comments or screenshots to attach if any -->

<!-- 
Don't forget to:

 ✓ Self-review your changes to make sure nothing unexpected slipped through
 ✓ Assign yourself to the PR
 ✓ Assign one or several reviewer(s)
 ✓ Once created, link this PR to its corresponding ticket
 ✓ Acknowledge any changes required to the Wiki
-->

Anviking committed:
Implement StakePoolLayer with listStakePools

fixup: use minBound :: Percentage

fixup: convert -> count

fixup whitespace spl

Add more detailed TODOs for detecting if we are unsynced

Fixup: accidental diff

Other approach

Anviking committed:
Connect StakePoolLayer with Api

fixup: f -> mkApiStakePool

fixup: error in bridge listStakePools

fixup: use withDBLayer

Other approach

KtorZ commented on lib/http-bridge/src/Cardano/Wallet/HttpBridge.hs:
```diff
@@ -172,6 +175,11 @@ serveWallet (cfg, tr) databaseDir listen bridge mAction = do
 
+    dummyPool :: StakePoolLayer IO
+    dummyPool = StakePoolLayer
+        { listStakePools = return []
```
--> `error "StakePoolLayer: not implemented for http-bridge"`

KtorZ commented on lib/jormungandr/src/Cardano/Wallet/Jormungandr.hs:
```diff
@@ -154,13 +163,17 @@ serveWallet (cfg, tr) databaseDir hostPref listen lj beforeMainLoop = do
     toWLBlock = J.convertBlock
 
+    poolDBPath :: Maybe FilePath -> Maybe FilePath
+    poolDBPath = fmap (<> "/stake-pool-metrics.sqlite")
```
`stake-pool-metrics.sqlite` --> `stake-pools`

Anviking replied:
I wouldn't mind `stake-pools.sqlite` too much, but why?

I chose `stake-pool-metrics.sqlite` to highlight that the purpose is not to simply "store stake-pools" (what does that even mean / why would we need to do that anyway?), but to collect metrics about them.

What do you think?

KtorZ replied:
Well, I'd rather be a bit more general with the file name here. I don't like to "code for the future", but for things that live at the boundaries of the system, we have to be a bit careful. "stake-pools" is a bit more general, and opens up for extension. In the end, there's no need to be that precise at this stage. 

As for the `.sqlite` extension, my bad, I thought this was added automatically by persistent. Still, I'd prefer to have the name for the database part of the `Pool.DB.Sqlite` module, very much like we do for the wallets databases.

Anviking replied:
Something factory-like felt overkill. But I moved this function to `Cardano.Pool.DB.Sqlite` (but without Maybe).

Anviking commented on lib/core/src/Cardano/Pool/Metrics.hs:
```diff
@@ -18,6 +18,11 @@ module Cardano.Pool.Metrics
     , ErrMetricsInconsistency (..)
 
+      -- * StakePoolLayer
+    , StakePoolLayer (..)
```
Maybe `StakePoolLayer` should live in its own module, not sure

KtorZ replied:
If there's no particular reason to have it in a separate module, then let's not for now.

KtorZ commented on lib/jormungandr/src/Cardano/Wallet/Jormungandr.hs:
```diff
@@ -134,7 +137,12 @@ serveWallet (cfg, tr) databaseDir hostPref listen lj beforeMainLoop = do
-            startServer tr nPort bp rndApi seqApi
+
+            -- StakePool
+            (_, spDB) <- PoolSqlite.newDBLayer cfg tr (poolDBPath databaseDir)
```
We should use `withDBLayer` here, actually I don't even know why we have a `newDBLayer` still exported and available... We always want the bracketing style to get proper resource handling.

Anviking replied:
Done, I don't know why, I expected it to be more complicated.

KtorZ replied:
I think it's actually even simpler... :sweat_smile:

KtorZ commented on lib/core/src/Cardano/Wallet/Api/Server.hs:
```diff
@@ -1432,3 +1449,12 @@ instance LiftHandler (Request, ServantErr) where
+         ErrMetricsIsUnsynced p ->
+             apiError err400 NotSynced $ mconcat
+                 [ "I can't list stake pools yet because I need to scan the "
+                 , "blockchain for metrics first. I'm at "
```
`for metrics` => `for data` :thinking: ?

Anviking replied:
Maybe `for data` could give the impression that we don't know which stake-pools exist without first scanning the blockchain, which is false. 🤔

KtorZ replied:
Yes. Although it conflicts with "I can't list stake pools" then. I mean I am neat-picking here ^^"

KtorZ commented on lib/core/src/Cardano/Wallet/Api/Server.hs:
```diff
@@ -721,14 +726,26 @@ postTransactionFee ctx (ApiT wid) body = do
-listPools _ctx = throwError err501
+listPools spl = liftHandler (map f <$> listStakePools spl)
+  where
+    f
```
`f` => `mkApiStakePool`

KtorZ commented on lib/core/src/Cardano/Pool/Metrics.hs:
```diff
@@ -64,12 +79,53 @@ activityForEpoch epoch s =
+                        Left e -> liftIO $ throwIO e
+                _ -> do
+                    let p = either (error . show) id $ mkPercentage (0::Int)
+                    throwE $ ErrMetricsIsUnsynced (Quantity p)
```
suggestion:

```hs
throwE $ ErrMetricsIsUnsynced (Quantity minBound)
```

:clap: avoid :clap: partial :clap: code :clap:

KtorZ commented on lib/core/src/Cardano/Pool/Metrics.hs:
```diff
@@ -64,12 +79,53 @@ activityForEpoch epoch s =
+                    throwE $ ErrMetricsIsUnsynced (Quantity p)
+        }
+  where
+    convert = Map.map (Quantity . fromIntegral . length)
```
`convert` => `count`

Anviking replied:
Nice one!

Anviking committed:
Actually start `monitorStakePools` on startup!

KtorZ commented on lib/core/src/Cardano/Wallet/Api/Server.hs:
```diff
@@ -375,12 +379,13 @@ coreApiServer
     :<|> wallets ctx
     :<|> transactions ctx
-    :<|> pools ctx
+    :<|> pools spl
```
What about moving `pools` to a separate tree and have something more like:

```hs
server = coreApiServer seqCtx :<|> compatibilityApiServer rndCtx seqCtx :<|> poolServer spl
```

Anviking replied:
Done in `Re-organize API: separate StakePoolApi from CoreApi`

KtorZ commented on lib/jormungandr/src/Cardano/Wallet/Jormungandr.hs:
```diff
@@ -134,7 +137,12 @@ serveWallet (cfg, tr) databaseDir hostPref listen lj beforeMainLoop = do
+
+            -- StakePool
+            PoolSqlite.withDBLayer cfg tr (poolDBPath databaseDir) $
+                \splDB -> do
```
why the line-break here :grimacing: ?

Anviking replied:
Because we have a rule saying lines must be 80 characters or less 😬 This one would have been 81.

Maybe importing `Cardano.Pool.DB.Sqlite` as `Pool` instead of `PoolSqlite` would actually be nice though…

KtorZ replied:
Hmm. Yes. But the guidelines are... Guidelines. When it makes sense, we can break them and we sometimes do.

Anviking committed:
Rename readCursor to readPoolProductionCursor

Anviking committed:
Re-organize API: separate StakePoolApi from CoreApi

KtorZ commented on lib/core/src/Cardano/Pool/Metrics.hs:
```diff
@@ -79,6 +90,74 @@ data Block = Block
+                case mDbTip of
+                    Nothing -> throwE $ ErrMetricsIsUnsynced (Quantity minBound)
+                    Just dbTip -> do
+                        liftIO $ logInfo tr $ mconcat
```
This is at most a  `debug` log line.

KtorZ commented on lib/core/src/Cardano/Pool/Metrics.hs:
```diff
@@ -79,6 +90,74 @@ data Block = Block
+                            , pretty dbTip
+                            ]
+                        let p = progress dbTip nodeTip
+                        throwE $ ErrMetricsIsUnsynced (Quantity p)
```
I'd suggest to move the whole `when` body here to a separate function to make the top-level one more readable

```hs
when (Map.null dist || Map.null prod) $
    computeProgress >>= throwE . ErrMetricsIsUnsynced
```

KtorZ commented on lib/core/src/Cardano/Pool/Metrics.hs:
```diff
@@ -79,6 +90,74 @@ data Block = Block
+        _ -> return Nothing
+
+    progress :: BlockHeader -> BlockHeader -> Percentage
+    progress tip target =
```
Why not use the progress calculation from the primitive types?

Anviking replied:
1. This `blockHeight`-calculation would return 100% when we are at the node tip. I.e. when the `stake-pool list` results become available. `syncProgress` compare to the current time. If the node tip is behind the current time, the progress might then jump from 50% to, show results.
2. This `blockHeight` approach is simple and gives us (seemingly) perfectly linear increase in progress.
3. `syncProgress` returns `SyncProgress = Ready | Restoring`, which we don't want.
4. `syncProgress` has a `tolerance` which we don't want. Here we _know_ that we are not in sync, and want to show the _exact_ progress to the user. Even if that would happen to be 99% or 100%.

KtorZ commented on lib/jormungandr/src/Cardano/Wallet/Jormungandr.hs:
```diff
@@ -138,7 +150,18 @@ serveWallet (cfg, tr) databaseDir hostPref listen lj beforeMainLoop = do
+            PoolSqlite.withDBLayer cfg tr (poolDBPath databaseDir) $
+                \splDB -> do
+
+                    -- TODO: I believe @KTorZ wants us to use @WorkerRegistry@
```
Not `WorkerRegistry`. There's no need for this. But `newWorker` would be good. With `void $ forkIO`, we are just firing a thread in the wild and we have no control over it..

Anviking replied:
Separate PR? And can we talk about it first or maybe you want to do do that?

KtorZ commented on lib/jormungandr/test/data/jormungandr/genesis.yaml:
```diff
@@ -20,14 +20,14 @@ blockchain_configuration:
 
   # The number of blocks (*10) per epoch
-  epoch_stability_depth: 10
+  epoch_stability_depth: 2
```
:thinking: should this be committed ?

Anviking replied:
No

Anviking replied:
(Unless we have an integration test that benefits from it)

KtorZ commented on lib/core/src/Cardano/Pool/DB/Sqlite/TH.hs:
```diff
@@ -44,12 +44,13 @@ share
     poolProductionPoolId         W.PoolId     sql=pool_id
-    poolProductionSlot           W.SlotId     sql=slot
+    poolProductionSlot           Word16       sql=slot
+    poolProductionEpoch          Word64       sql=epoch
```
Why here? We have to be careful with this.

Anviking replied:
👍 It turns out this is in fact not needed. Dropped.

Anviking committed:
Fixup in DB

Anviking commented on lib/core/src/Cardano/Pool/DB/Sqlite.hs:
```diff
@@ -168,7 +178,8 @@ selectPoolProduction
 selectPoolProduction epoch = fmap entityVal <$> selectList
-    [PoolProductionSlot >=. SlotId epoch 0]
+    [ PoolProductionSlot >=. SlotId epoch 0
+    , PoolProductionSlot <. SlotId (epoch + 1) 0 ]
```
Added a quick fixup for this, but no test.

KtorZ committed:
rename default file for pools db to use plural form

KtorZ reviewed (approved):
LGTM. Will tackle remaining points in subsequent PRs.

Anviking committed:
Fixup weeder

KtorZ committed:
review HTTP error code for stake pools endpoint and integration spec assertions

NOTE: 4xx code are for CLIENTS errors. Not being able to answer because we aren't synced is not a client error.

KtorZ committed:
add an initial log line when starting the stake pool monitoring

KtorZ committed:
lower down API logging severity for 503 errors

KtorZ committed:
slightly review retry logic in chain follower to use an exponential backoff in 'nextBlocks' too

paweljakubas reviewed (approved):
Nice work!

Anviking commented on lib/jormungandr/test/integration/Test/Integration/Jormungandr/Scenario/CLI/StakePools.hs:
```diff
@@ -18,16 +18,14 @@ import System.Exit
-            \situation. Here's some information about what happened: \n"
-        o `shouldBe` mempty
-        c `shouldBe` ExitFailure 1
+        eventually $ do
```
😱 I was wondering if there was something as simple as `eventually`

KtorZ replied:
:man-shrugging:
